src/main/java/org/subethamail/smtp/server/SourceTrackingSessionLifecycleListener.java

```diff
@@ -0,0 +1,140 @@
+
+    /** Session drop response */
+    private final SessionStartResult drop;
+
+    private final int maxConnectionsPerSource;
+    private final ConcurrentMap<SrcKey, Integer> counts;
+
+    /**
+     * Create a new {@link SourceTrackingSessionLifecycleListener} with default reject message:
+     * {@code "421 Too many connections, try again later"}.
+     *
+     * @param maxConnectionsPerSource maximum number of concurrent connection per remote source ip
+     */
+    public SourceTrackingSessionLifecycleListener(int maxConnectionsPerSource) {
+        this(maxConnectionsPerSource, 421, "Too many connections, try again later");
+    }
+
+    /**
+     * Create a new {@link SourceTrackingSessionLifecycleListener} with custom reject message
+     * @param maxConnectionsPerSource maximum number of concurrent connection per remote source ip
+     * @param code SMTP code
+     * @param message SMTP message
+     */
+    public SourceTrackingSessionLifecycleListener(int maxConnectionsPerSource, int code, String message) {
+        super();
+
+        this.maxConnectionsPerSource = maxConnectionsPerSource;
+        this.drop = SessionStartResult.failure(code, message);
+
+
+        this.counts = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public SessionStartResult onSessionStart(Session session) {
+        try {
+            counts.compute(toKey(session), (k, v) -> {
+                if (v == null) {
+                    return 1;
+                } else {
+
+                    if (v == maxConnectionsPerSource) {
+                        throw LimitReachedException.INSTANCE;
+                    } else {
+                        return ++v;
+                    }
+                }
+            });
+        } catch (LimitReachedException limit) {
+            return drop;
+        }
+        return SessionStartResult.success();
+    }
+
+    @Override
+    public void onSessionEnd(Session session) {
+        counts.compute(toKey(session), (k, v) -> {
+            if (--v == 0) {
+                return null;
+            } else {
+                return v;
+            }
+        });
+    }
+
+    private static SrcKey toKey(Session session) {
+        final byte[] src = session.getSocket().getInetAddress().getAddress();
+        return new SrcKey(src);
+    }
+
+    /**
+     * A lightweight exception to avoid to create non useful stacktraces, this exception is use only internally
+     * and never leaked out this class.
+     */
+    @SuppressWarnings("serial")
+    private static final class LimitReachedException extends RuntimeException {
+        public static final LimitReachedException INSTANCE = new LimitReachedException();
+
+        public LimitReachedException() {
+            /* Disables stacktraces and suppressions */
+            super("Limit reached", null, false, false);
+        }
+    }
+
+    /**
+     * Just a wrapper around a byte array to be used as key for a map.
+     */
+    private static final class SrcKey {
```

Comment:
You don't need this class `SrcKey` because `InetAddress` already implements `equals` and `hashCode` based on `getAddress()`. See https://docs.oracle.com/javase/8/docs/api/java/net/InetAddress.html#equals-java.lang.Object-. As a consequence your `counts` map can be `ConcurrentMap<InetAddress, Integer>`.